drivers/ethernet/eth_stm32_hal.c

```diff
@@ -439,7 +562,18 @@ static struct net_pkt *eth_rx(const struct device *dev, uint16_t *vlan_tag)
 
 	heth = &dev_data->heth;
 
-#if defined(CONFIG_SOC_SERIES_STM32H7X)
+#if defined(CONFIG_ETH_STM32_HAL_API_V2)
+	if (HAL_ETH_ReadData(heth, &appbuf) != HAL_OK) {
+		/* no frame available */
+		return NULL;
+	}
+
```

Comment:
By looking in the disassembly, it seems that gcc 10.3.1 optimizes away the case that a packet has been received. I suspect that this is due to some undefined behaviour (UB). Compiling with `-fno-strict-aliasing` does not change anything, but with `-Og` it seems to be fine (as `eth_rx` is not inlined)